### PR TITLE
HDFDataset: add get_data_by_seq_tag

### PR DESCRIPTION
**Pull request by vieting**
For a use case that I have in mind, I need to be able to get data from an `HDFDataset` based on the sequence tag. I added `HDFDataset.get_data_by_seq_tag()` for that purpose.

It is very similar to `HDFDataset.get_data()` which has a special case `if self.cache_byte_size_total_limit > 0`. I'm not familiar with that and not sure how to handle this here, that's why I mark this PR as draft.

### CODE REVIEW

File: returnn/datasets/hdf.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ def __init__(self, files=None, use_cache_manager=False, **kwargs):
     self.data_dtype = {}  # type: typing.Dict[str,str]
     self.data_sparse = {}  # type: typing.Dict[str,bool]
     self._num_codesteps = None  # type: typing.Optional[typing.List[int]]  # accumulated sequence length per target
+    self._tag_to_real_idx_map = None  # type: typing.Optional[typing.Dict[str,int]]
```
Comment (albertz) [inline returnn/datasets/hdf.py:51]: Why do you introduce that? You already have `_tag_idx` which is exactly the same?

File: returnn/datasets/hdf.py [hunk 2]
```diff
@@ -355,6 +377,33 @@ def _get_tag_by_real_idx(self, real_seq_idx):
...
+    """
+    :param str seq_tag:
+    :rtype: int
+    """
+    if self._tag_to_real_idx_map is None:
+      self._build_tag_to_real_idx_map()
+
+    return self._tag_to_real_idx_map[seq_tag]
+
+  def _build_tag_to_real_idx_map(self):
```
Comment (albertz) [inline returnn/datasets/hdf.py:390]: You don't need that. Instead, there is already `_update_tag_idx`.

File: returnn/datasets/hdf.py [hunk 3]
```diff
@@ -355,6 +377,33 @@ def _get_tag_by_real_idx(self, real_seq_idx):
     s = self._decode(s)
     return s
 
+  def _get_real_idx_by_tag(self, seq_tag):
```
Comment (albertz) [inline returnn/datasets/hdf.py:380]: I don't think such a helper function is needed. Just call `_update_tag_idx` and access `_tag_idx` directly.

File: returnn/datasets/hdf.py [hunk 4]
```diff
@@ -273,6 +274,27 @@ def get_data(self, seq_idx, key):
...
+    return self._get_data_by_real_seq_idx(real_seq_idx, key)
+
+  def get_data_by_seq_tag(self, seq_tag, key):
+    """
+    :param str seq_tag:
+    :param str key:
+    :rtype: numpy.ndarray
+    """
+    if self.cache_byte_size_total_limit > 0:  # Use the cache?
+      assert False  # what should we do here?
```
Comment (albertz) [inline returnn/datasets/hdf.py:286]: Throw a better exception instead, like `raise Exception("%s: get_data_by_seq_tag not supported with cache" % self)` or so.

**Review by albertz — APPROVED** [review]
Looks ok to me. Maybe either @JackTemaki or @patrick-wilken should check as well.

**Review by JackTemaki — APPROVED** [review]

**Comment by patrick-wilken** [comment]
Nothing wrong with the code. But if we in general want to support getting data via tag this maybe should become part of the Dataset base class at some point.
And just as a side note: I intentionally removed calls to `self._update_tag_idx()` in the past to only create `self._tag_idx` if really needed, because it can get large if the number of sequences is huge. But that does not seem to be the case for you.

**Comment by albertz** [comment]
> Nothing wrong with the code. But if we in general want to support getting data via tag this maybe should become part of the Dataset base class at some point.

Yea, we thought about this. But basically almost every dataset would need a custom implementation, and this implementation can be non-straightforward, or even tricky (e.g. `ExternSprintDataset`), or also impossible.

Also, this is a function which would not be used by RETURNN itself but only by user code. In this case, it is used to implement a new custom dataset which wants random access by seq tag to another (HDF) dataset.

> And just as a side note: I intentionally removed calls to `self._update_tag_idx()` in the past to only create `self._tag_idx` if really needed

But we do the same here? It is only called when really needed.

**Pull request merged**